keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,436 @@
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+#### Fast/Reliable (but Non-Source-IP-Preserving) LoadBalancer Services
+
+As an application developer, I want incoming service traffic from
+outside the cluster to be delivered to my pods as efficiently as
+possible (as with `Local` traffic policy), but I don't care if the
+source IP is preserved, and I _do_ care about maximizing reliability
+when pods restart or move between nodes (which can cause hiccups with
+`externalTrafficPolicy: Local` in some cases).
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
+topology and zone-level topology. Another possibility would be to have
+a new annotation for node-level topology.)
```

Comment:
Based on my comment above, I think the best approach is to integrate this more closely with topology hints because distributions may need to change depending on how endpoints are distributed. Having separate paths in EndpointSlice controller for either approach could get rather complicated.